openssl/src/ts.rs

```diff
@@ -0,0 +1,457 @@
+//! Partial interface to OpenSSL Time-Stamp Protocol (RFC 3161) implementation.
+//!
+//! This module provides a partial interface to OpenSSL's TSP implementation.
+//! The aim is to provide enough functionality for a client to request and
+//! verify timestamps returned by a Time Stamp Authority.
+use foreign_types::{ForeignType, ForeignTypeRef};
+use libc::{c_int, c_long, c_uint};
+
+use std::ptr;
+
+use asn1::{Asn1IntegerRef, Asn1ObjectRef};
+use error::ErrorStack;
+use ffi;
+use hash::MessageDigest;
+use pkey::{HasPrivate, PKeyRef};
+use x509::{X509Algorithm, X509Ref};
+use {cvt, cvt_p};
+
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::TS_MSG_IMPRINT;
+    fn drop = ffi::TS_MSG_IMPRINT_free;
+
+    /// A message imprint contains the hash of the data to be timestamped.
+    pub struct TsMsgImprint;
+
+    /// Reference to `TsMsgImprint`.
+    pub struct TsMsgImprintRef;
+}
+
+/// A builder used to construct a `TsMsgImprint`
+pub struct TsMsgImprintBuilder(TsMsgImprint);
+
+impl TsMsgImprintBuilder {
```

Comment:
TS_MSG_IMPRINT isn't reference counted, so there's no need for a separate builder type.

Reply:
Fixed, thanks!